### PR TITLE
Adding messages

### PR DESCRIPTION
**Pull request by jrmlhermitte**
Should we allow messages to pass? I think something like this would be useful.

For example here, we send a clear message which should clear various elements of the stream that hold state. Right now, only this is performed:
```python
    s = Stream()
    # increment
    def acc1(x1, x2):
        return x1 + x2
    from streamz.core import ClearMSG

    cc = ClearMSG()

    sout = s.accumulate(acc1)

    Lout = sout.sink_to_list()

    s.emit(1)
    s.emit(2)
    s.emit(cc)
    s.emit(3)
    assert Lout == [1, 3, 3]
```

but I'd find this useful. One could change these messages to node specific, like "CLEAR_ACC", or some control stream with flags like: `Clear(clear_acc=True,...)`.

any thoughts? This is meant to open discussion. I find it more convenient than an issue in this case.
The various messages or message handler can perhaps also be handled by dispatching, to allow for more convenient subclassing.

This was a quick thought. If it's of interest in the base class we can discuss adding more.

### CODE REVIEW

**Comment by mrocklin** [comment]
Do other systems do this?  If so then what are their semantics?

The alternative here would be to ask the user to implement this sort of logic within their operations, either in their `acc1` function or as a separate stream operation.

File: streamz/tests/test_core.py [hunk 1]
```diff
@@ -931,3 +931,28 @@ def slow_write(x):
...
+    s = Stream()
+    # increment
+    def acc1(x1, x2):
+        return x1 + x2
+    from streamz.core import ClearMSG, IgnoreMSG
+
+    clear_msg = ClearMSG()
+    ignore_msg = IgnoreMSG()
+
+    sout = s.accumulate(acc1)
```
Comment (mrocklin) [inline streamz/tests/test_core.py]: What would happen if you had a map before the accumulate?  Does map (or all other stream operations) also need to know what to do with ClearMSGs?

Comment (CJ-Wright) [reply streamz/tests/test_core.py]: I would imagine that all `update` methods will need a cutout for messages.

**Comment by CJ-Wright** [comment]
I'm not certain that we can ask the user to modify everything, as there are some cases where we buffer information without user input. For example `zip_latest`. In the case of `zip_latest` you may decide that all the information in the zip_latest buffer is no longer applicable and you want it to go back into "buffer my main stream until my secondary stream emits" mode. A clear message could be helpful here.

**Comment by jrmlhermitte** [comment]
I will have to get back to you about this, sorry busy.
I spent a few min looking at ReactiveX.
They split `emit` into three different functions: `onNext` (regular emit), `onError` and `onCompleted`.
The way error handling works is that when an error is received, the node 'disconnects' (`unsubscribe`s) from its parent stream and the error is passed through to the other elements in the stream. This effectively stops data from passing through the stream graph.

If you want to prevent this behaviour from happening, you must insert a [`Retry`](http://reactivex.io/documentation/operators/retry.html) node in between the node that could return an error and the node that would receive an error. Upon receiving an error, this node will not pass it through and resubscribe to its parent stream.

I don't like this method but I do like the splitting of `emit` into different methods. We could have `emit` dispatch the element to something like `onNext`, `onCompleted`, `onClear` and `onError`. All accumulators, zips etc would then have to implement their own `onClear` methods for example. The default for `onClear` would be `onNext` (`emit`).

Thoughts? This ties in with #86 as well I believe.

**Comment by jrmlhermitte** [comment]
(Just for extra reference, @CJ-Wright has already implemented something similar on the `update` level [here](https://github.com/xpdAcq/SHED/blob/master/shed/event_streams.py#L271). `dispatch` will dispatch to a `start`, `event`, `descriptor` and `stop` methods which intercept different types of messages/documents.)

File: streamz/core.py [hunk 2]
```diff
@@ -28,6 +28,16 @@
 logger = logging.getLogger(__name__)
 
 
+class ClearMSG:
```
Comment (CJ-Wright) [inline streamz/core.py]: You may want a MSG base class so you can just check if an item is an instance if that class

**Comment by martindurant** [comment]
superceded by #249 and its discussion, I think

**Pull request closed without merging**